keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,399 @@
+- GCE: here the risk is valid once the user exceeds the threshold of 5 forwarding rules.
+- IBM Cloud: no risk
+- OpenStack: here the risk is that there is almost no chance to analyze all the public OpenStack cloud providers with regard to their pricing policies
+- Oracle: no risk
+- Tencent Cloud: no risk
+
+#### API change and upgrade/downgrade situations
+
+We relax here a validation rule, which is considered as an API-breaking act by the [K8s API change guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) Even if the change is implemented behind a feature flag the following actions can cause problems:
+- the user creates a Service with mixed protocols and then
+  - turns off the feature flag; or
+  - executes K8s version rollback to the N-1 version where this feature is not available at all
+
+When investigating the possible issues with such a change we must consider [the supported version skew among components](https://kubernetes.io/docs/setup/release/version-skew-policy/), which are the K8s API server and the cloud controller managers (CPI implementations) in our case.
+
+First of all, feature gate based (a.k.a conditional) field validation must be implemented as defined in the [API change guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version) One can see a good example for a case when an existing field got a new optional value [here](https://github.com/kubernetes/kubernetes/issues/72651). This practice ensures that upgrade and rollback between _future releases_ is safe. Also it enables the further management of existing API objects that were created before turning off the feature flag. Though it does not save us when a K8s API server version rollback is executed to a release in which this feature is not implemented at all.
+
+Our feature does not introduce new values or new fields. It enables the usage of an existing value in existing fields, but with a different logic. I.e. if someone creates a Service with mixed protocol setup and then rollbacks the API server to a version that does not implement this feature the clients will still get the Service with mixed protocols when they read that via the rollback'd API. If the client (CPI implementation) has been rollbacked, too, then the client may receive such a Service setup that it does not support.
+
+- Alibaba: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. If this feature is enabled in an API server and then the API server rollback is executed the CPI can still handle the Services with mixed protocol sets.
+- AWS: no risk. The current CPI accepts Services with TCP protocol only, i.e. after a K8s upgrade a user still cannot use this feature. Consequently, a rollback in the K8s version does not introduce any issues.
```

Comment:
https://github.com/kubernetes/kubernetes/pull/92109 will introduce UDP support to the AWS CPI. It's been approved, and will hopefully be merged soon. The updated CPI will _not_ work correctly with mixed TCP/UDP without further work, c.f. https://github.com/kubernetes/kubernetes/pull/92109#discussion_r439730341

Reply:
@TBBle Fixed the AWS related parts.